### DISCUSSION TITLE
Behaviour Modify StateProvider.autoDispose  state using ref.read() & ref.watch()

### DISCUSSION THREAD

**Discussion by zgramming** [rrousselGit/riverpod#1401] [Q&A]
Hello Remi and all maintaners,

I curious about `StateProvider.autoDispose` behaviour when using `ref.read` & `ref.watch`. 

In my little example, I have a form in which it has 2 inputs **Name** & **Email** and button submit to obtain input in form.

In `onChange` each input, i update `form state` depend of input value using 2 option : 
1. `ref.read(formParameterState.notifier).update((state)=> state.copyWith(name: val))`
2. `ref.watch(formParameterState.notifier).update((state) => state.copyWith(name: val));`

The different about 2 option above is only use `read` or `watch`. But strangely when using the first option, `form state` autoDispose is trigger,  otherwise when using the second option, it doesn't trigger autoDispose.

This behavior also appears when I want to obtain `form state` when click button, when using `ref.read(formParameterState)` is trigger Autodispose, otherwise when using `ref.watch(formParameterState.notifier).state` is doesn't trigger autodispose.

For better imaging my case, this is gif when using `ref.read` and `ref.watch`

## ref.read (trigger autodispose)
![stateprovider_read](https://user-images.githubusercontent.com/38829404/167764940-bd41383c-e971-446e-9379-073cf8eddad8.gif)

## ref.watch (not trigger autodispose)
![stateprovider_watch](https://user-images.githubusercontent.com/38829404/167764951-02e70490-fb35-4d88-a734-8d705c7b62c3.gif)

My question is, this is how stateProvider.autodispose works  ? 
Thank's 😁
 
## Repo 

[Repository](https://github.com/zgramming/state-provider-autodispose-behaviour)

## Sample source code 

```
import 'dart:developer';

import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

class FormParameterState extends Equatable {
  const FormParameterState({
    this.name = '',
    this.email = '',
  });
  final String name;
  final String email;

  @override
  List<Object> get props => [name, email];

  @override
  String toString() => 'FormParameterState(name: $name, email: $email)';

  FormParameterState copyWith({
    String? name,
    String? email,
  }) {
    return FormParameterState(
      name: name ?? this.name,
      email: email ?? this.email,
    );
  }
}

final formParameterState = StateProvider.autoDispose(
  (ref) {
    ref.onDispose(() {
      log("Trigger Dispose [formParameterState]");
    });
    return const FormParameterState();
  },
);

void main() {
  runApp(
    const ProviderScope(
      child: MyApp(),
    ),
  );
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const WelcomePage(),
    );
  }
}

class WelcomePage extends ConsumerStatefulWidget {
  const WelcomePage({Key? key}) : super(key: key);

  @override
  _WelcomePageState createState() => _WelcomePageState();
}

class _WelcomePageState extends ConsumerState<WelcomePage> {
  final nameController = TextEditingController();
  final emailController = TextEditingController();

  @override
  void dispose() {
    nameController.dispose();
    emailController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Form"),
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Expanded(
            child: SingleChildScrollView(
              child: Form(
                child: Padding(
                  padding: const EdgeInsets.all(16.0),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      TextFormField(
                        controller: nameController,
                        decoration: const InputDecoration(
                          border: OutlineInputBorder(),
                          hintText: "Name",
                        ),
                        onChanged: (val) {
                          // ref
                          //     .read(formParameterState.notifier)
                          //     .update((state) => state.copyWith(name: val));
                          ref.watch(formParameterState.notifier).update((state) => state.copyWith(name: val));
                        },
                      ),
                      const SizedBox(height: 20),
                      TextFormField(
                        controller: emailController,
                        decoration: const InputDecoration(
                          border: OutlineInputBorder(),
                          hintText: "Email",
                        ),
                        onChanged: (val) {
                          // ref
                          //     .read(formParameterState.notifier)
                          //     .update((state) => state.copyWith(email: val));
                          ref
                              .watch(formParameterState.notifier)
                              .update((state) => state.copyWith(email: val));
                        },
                      ),
                      const SizedBox(height: 20),
                    ],
                  ),
                ),
              ),
            ),
          ),
          ElevatedButton(
            onPressed: () {
              // final formRead = ref.read(formParameterState);
              // log("[Form Parameter Read] Name : ${formRead.name}");
              // log("[Form Parameter Read] Name : ${formRead.email}");

              final formWatch = ref.watch(formParameterState.notifier).state;
              log("[Form Parameter Watch] Name : ${formWatch.name}");
              log("[Form Parameter Watch] Name : ${formWatch.email}");
            },
            child: const Text("sss"),
          )
        ],
      ),
    );
  }
}

```

**Comment by lucavenir** [comment]
Hi,

Here you wrote:
> The different about 2 option above is only use `read` or `watch`. But strangely when using the first option, `form state` autoDispose is trigger, otherwise when using the second option, it doesn't trigger autoDispose.

You actually wrote exactly why the `dispose` callback is not triggered.
Let's read the `.autoDispose` modifier's [definition](https://riverpod.dev/docs/concepts/modifiers/auto_dispose/)
> ```final userProvider = StreamProvider.autoDispose<User>((ref) { /* something */  });```
> That's it. Now, the state of `userProvider` will automatically be destroyed when it is no longer used.

I guess this sentence has room for improvement, but let me rephrase that better for your context:
> _That's it. Now, this `Provider` will automatically be `disposed` whenever **no one watches/listens to it anymore**._


In other words, if you're using `ref.watch(someStateNotifierProvider.notifier).update(...)`, you're asking Riverpod to _actively_ `watch` this variable; therefore, by definition, as long as this provider is being watched or listened to, it _won't_ be disposed.

It's usually a good practice to use `read` to access the `StateNotifier` of a `StateNotifierProvider`, whereas actively watching it leads to some unexpected behavior. Commonly, `StateNotifier` centralizes logic and has methods, getters and setters to handle its internal state. If you want to `watch` its `state`, just `watch` its provider.

On the other hand, "watching its logic handler" doesn't make much sense to me, but maybe it's just me.

@rrousselGit Correct me if I'm wrong. Maybe it can be a good idea to document the "common pitfalls" of Riverpod (I'd like to contribute, I've done so many mistakes with it).

**Reply by zgramming** [reply]
Thank's for your reply. @lucavenir  

But I still don't get the meaning of this behaviour, using `ref.read(X.notifier).update()` i only want to update / modify old state to new state, and i expected not to dispose or lost the state. When using `ref.watch(X.notifier).update()` the result is as i expected.

Because using `ref.read(X.notifier).update()` will make state lost every `onChange` trigger.

So is in my case it's really supposed to use `ref.watch(X.notifier).update()` instead of `ref.read(X.notifier).update()` ? 
In real project, i usually often using `autoDispose` modifier in my `StateProvider`, Because i want the state is dispose when the state is not used.

**Reply by lucavenir** [reply]
Hi again. I checked your repo and I see that indeed it's weird at first, but after a while I managed to understand what's going on! Luckily this makes total sense, although I am left with some doubts (see the end). But I fixed your problem. Let's analyze this.

In case two we just `watch` providers, via `ref.watch(provider.notifier).update(...)`. In this case, the provider is _never_ disposed, but that's intended! As explained above, as long as you `watch` it, the provider won't dispose itself. At this point, this is clear to me. You're not supposed to `watch` just to _alter_ state. See previous message.

In case one we _instead_ use `ref.read` everywhere.
I can see the provider is disposed everytime we type something, and not just that: your provider _is being rebuilt_ every time we type a letter and then immediately disposed after a few milliseconds. At first this is weird and unexpected, but let's think this through.

Here's what happens: since `formParameterState` is never watched, at the first letter we type, it's being computed. This is supposed to happen: it's the first time we _use it_ (altering its state).

Then, _again_, since no one is actively listening (or watching) to it, it's being disposed. This is also supposed to happen: no watchers = dispose it.

Then again at the next letter type you're reading this disposed provider to update it. At first I would expect this situation to `throw` an error (reading a disposed provider is not allowed), but this is not happening _since you're asking the provider to expose its notifier to update it_. Therefore, the provider is computed again and then, _again_, it is disposed since no one listens to it.

Finally, when you click your button, on `ref.read(formParameterState.notifier).read` *your provider is rebuilding again* (therefore it doesn't expose what you've typed so far!), and then disposing again (no one listens to it).

One could just say: let's use `ref.watch(formParameterState)` in our `onPressed` callback. But that won't work, since as the docs say, `watch` behaves incorrectly inside buttons callbacks (and such).

So - how to solve this? You could use `ref.listen` in the `build` method of `_WelcomePageState`. Or maybe you can wrap your `ElevatedButton` with a `Consumer`, although this will rebuild the button everytime we type a letter.

Wrapping this up, we should indeed use Riverpod like this:
- When we need to _alter_ state, just use `ref.read` onto the notifier;
- When we need to _use_ state, just use `ref.watch` onto the provider itself;
- BUT if you want to _use_ state in a Button's callback (or similar), keep in mind that `watch` won't work as expected.

These situations may be a "common pitfall" that should be documented in my opinion. This discussion immediately made me think: why isn't using `ref.read(formParameterState)` (in the Button callback) throwing an error, since that provider is in a `disposed` state? @rrousselGit maybe you can help me figure out this last question?

Anyways, I hope this helps.

**Reply by lucavenir** [reply]
I just created a PR to your repo, so that you can see the instant fix

**Reply by zgramming** [reply]
@lucavenir Already see your PR, the result is as expected. But with this solution, make me thinking `what a waste make ref.listen(notifierX) listen to the stateProvider.autoDispose and prevent disposing the state`, even though I know the reason why it has to be like this like your statement : 
>Then, again, since no one is actively listening (or watching) to it, it's being disposed. This is also supposed to happen: no watchers = dispose it.

Isn't this a strange behavior of `StateProvider.autoDispose`, do you think the same as me ?

**Reply by zgramming** [reply]
@rrousselGit If you like , would you like to give some suggestion about this case ?
If @lucavenir statement is right and the solution should be like that (when using stateProvider.autoDispose, we consider make `ref.listen(notifierX)` to listen the provider and prevent disposed the state), I'm okay with it.

Thank's before 😁

**Reply by lucavenir** [reply]
> Isn't this a strange behavior of `StateProvider.autoDispose`, do you think the same as me ?

No, this is the intended behavior: if you're not watching or listening to it, the provider _will_ be disposed. There will be more API in the future (see 2.0), like `onCancel` to behave differently, but chances are that you can solve this without that.

Of course it feels weird to force a listen in the build method, that was just a quick fix as a proof of concept. Chances are that you need that value somewhere else, and therefore you'll use `Consumer` or something similar, there. It really depends on your use case. In this particular example there is no actual need of using `autoDispose`, so by just watching your snippet, one could simply remove that clause to make this work.

**Reply by zgramming** [reply]
When using `StateProvider.autoDispose` only for small cases like the example above, yes it will useless. Actually we are enough with the usual `StateProvider`.

But in real application like we `Navigator.push` to form page have many textfield and we need obtain the state when press button. After finish, we `pop` to previous page  and when we no longer need `state` we should dispose it automatic, This task is suitable for `state provider autodispose`.

But, when we use only `StateProvider` without auto dispose, we should take care dispose it manually when `pop` the page.

**Reply by zgramming** [reply]
@lucavenir By the way, thank you for taking the time to share your thoughts on my concerns 😀

**Reply by lucavenir** [reply]
I totally see your use case.
This what I'd do if I were you:

Alternative 1.
- Use a `StateProvider`, no `autoDispose` clause
- When invoking a `pop` function, clean up the state (e.g. `ref.read(provider.notifier).state = 0`)
- If you're worried about memory leaks, it's just a small notifier that cashes a "zero" value, that won't be a problem

Alternative 2.
- Use the `autoDispose` clause
- Either use the `listen` workaround or just use whatever you're inputting as a label somewhere with a `watch`. Something like:

```dart
Consumer(
  builder: (context, ref, _) {
    final label = ref.watch(yourAutoDisposeProvider);
    return Text(label);
  },
),
```
